### PR TITLE
std.format: Add overview table of all combinations of types and format characters.

### PR DESCRIPTION
**Pull request by berni44**
Final spurt: After this there will be only one more PR before the document format marathon ends. :-)

As this table has a normative touch, repetitions are somewhat intentional.

I also fixed two nits that I overlooked in the previous PR.

### CODE REVIEW

File: std/format/package.d [hunk 1]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
 precision) and $(I separator) by adding a number and a $(B
 '$(DOLLAR)') after the $(B '*').
 
+$(SECTION4 Types)
+
+The following table contains detailed descriptions for every
+available combination of type and format character. To avoid
+unnecessary repetition, here are some notes that apply to all or
+several entries:
```
Comment (RazvanN7) [inline std/format/package.d:277]: I would reorganize this paragraph the following way:

- start by explaining what this section does. E.g: "This section describes the result of combining types with format characters. It is organized in 2 subsections: a list of general information regarding the formatting of types in the presence of format characters and a table that contains details for every available combination of type and format character".
- continue with the general description paragraph: "When formatting types, the following rules apply:".
- finish with the table: "This table contains descriptions for every possible combination of type and format character"

File: std/format/package.d [hunk 2]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+
+$(UL
+  $(LI If the format character is upper case, the outcome is identical
+       to the lower case version, but uses upper case letters.)
```
Comment (RazvanN7) [inline std/format/package.d:281]: "If the format character is upper case, the resulting string will be formatted using upper case letters"

File: std/format/package.d [hunk 3]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+  $(LI The default precision for floating point numbers is 6 digits.)
+  $(LI Floating point numbers are rounded according to the
+       current rounding mode.)
```
Comment (RazvanN7) [inline std/format/package.d:284]: What is the current rounding mode?

File: std/format/package.d [hunk 4]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+  $(LI The floating point values `NaN` and `Infinity` are formatted as
+       `nan` and `inf`, possibly preceded by a sign.)
```
Comment (RazvanN7) [inline std/format/package.d:286]:
```suggestion
       `nan` and `inf`, possibly preceded by a  $(B +) or $(B -) sign.)
```

File: std/format/package.d [hunk 5]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+   $(TR $(MULTIROW_CELL 3, $(I Integral))
+        $(TD $(B 's'), $(B 'd'))
+            $(TD A signed decimal number. The $(B '#') flag is ignored.)
+   )
+   $(TR $(TD $(B 'b'), $(B 'o'), $(B 'u'), $(B 'x'), $(B 'X'))
+            $(TD An unsigned binary, decimal, octal or hexadecimal number.
+
+                 In case of $(B 'o') and $(B 'x'), the $(B '#') flag
+                 denotes that the number must be preceded by `0` and `0x`, with
+                 the exception of the value 0, where this not applies. For
```
Comment (RazvanN7) [inline std/format/package.d:330]:
```suggestion
                 the exception of the value 0, where this does not apply. For
```

File: std/format/package.d [hunk 6]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+                 fractional digits.
+
+                 When there are no fractional digits and the $(B '#') flag
+                 is $(I not) present, the dot is omitted.
+
+                 $(I Please note: the difference between $(B 'f') and $(B 'F')
+                 is only visible for `NaN` and `Infinity`.))
+   )
+   $(TR $(TD $(B 's'), $(B 'g'), $(B 'G'))
+            $(TD Short notation: If the absolute value is larger as `10 ^^ precision`
```
Comment (RazvanN7) [inline std/format/package.d:358]:
```suggestion
            $(TD Short notation: If the absolute value is larger than `10 ^^ precision`
```

File: std/format/package.d [hunk 7]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+                 Inside of a compound indicator the string is surrounded by double quotes
+                 and non printable characters are escaped. This can be avoided
+                 by preceding the compound indicator with a $(B '-') flag
+                 $(LPAREN)e.g. `"%-$(LPAREN)%s%$(RPAREN)"`$(RPAREN).)
+   )
+   $(TR $(TD $(B 'r'))
+            $(TD The sequence of characters, each formatted with $(B 'r').)
+   )
+   $(TR $(TD compound)
+            $(TD Like an array of characters.)
```
Comment (RazvanN7) [inline std/format/package.d:409]: Doesn't "As an array of characters" sound better than "Like an array of characters"? I, personally, don't like the latter. If you agree, please replace all such occurrences with "as".

File: std/format/package.d [hunk 8]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+   )
+   $(TR $(MULTIROW_CELL 3, $(I Array))
+        $(TD $(B 's'))
+            $(TD When the elements are characters, the array is formatted like
```
Comment (RazvanN7) [inline std/format/package.d:413]:
```suggestion
            $(TD When the elements are characters, the array is formatted as
```

File: std/format/package.d [hunk 9]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+   $(TR $(TD $(B 'r'))
+            $(TD The sequence of the elements, each formatted with $(B 'r').)
+   )
+   $(TR $(TD compound)
+            $(TD The sequence of the elements, each formatted according to the specifications
+                 given inside of the compound specifier.)
+   )
+   $(TR $(MULTIROW_CELL 2, $(I Associative Array))
+        $(TD $(B 's'))
+            $(TD As a sequence of the elements in unpredictable order. The output is
```
Comment (RazvanN7) [inline std/format/package.d:428]: I see that you used "as" here. Sounds so much better.

File: std/format/package.d [hunk 10]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+            $(TD The sequence of the elements, each formatted with $(B 'r').)
+   )
+   $(TR $(TD compound)
+            $(TD The sequence of the elements, each formatted according to the specifications
+                 given inside of the compound specifier.)
+   )
+   $(TR $(MULTIROW_CELL 1, $(I Struct))
+        $(TD $(B 's'))
+            $(TD When the struct has neither an applicable `toString`
+                 nor is an input range, it is formatted like `Type(field1, field2, ...)`.)
```
Comment (RazvanN7) [inline std/format/package.d:464]:
```suggestion
                 nor is an input range, it is formatted as follows: `StructType(field1, field2, ...)`.)
```

File: std/format/package.d [hunk 11]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+   $(TR $(MULTIROW_CELL 1, $(I Class))
+        $(TD $(B 's'))
+            $(TD When the class has neither an applicable `toString`
+                 nor is an input range, it is formatted as the
+                 fully qualified name of the class.)
+   )
+   $(TR $(MULTIROW_CELL 1, $(I Union))
+        $(TD $(B 's'))
+            $(TD When the union has neither an applicable `toString`
+                 nor is an input range, it is formatted as it's base name.)
```
Comment (RazvanN7) [inline std/format/package.d:475]:
```suggestion
                 nor is an input range, it is formatted as its base name.)
```

File: std/format/package.d [hunk 12]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+   )
+   $(TR $(MULTIROW_CELL 2, $(I Pointer))
+        $(TD $(B 's'))
+            $(TD A null pointer is formatted as 'null'. All other pointers are
+                 formatted as hexadecimal numbers with format character $(B 'X').)
```
Comment (RazvanN7) [inline std/format/package.d:480]:
```suggestion
                 formatted as hexadecimal numbers with the format character $(B 'X').)
```

File: std/format/package.d [hunk 13]
```diff
@@ -269,119 +269,242 @@ It's also possible to use positional arguments for $(I width), $(I
...
+   $(TR $(TD compound)
+            $(TD The sequence of the elements, each formatted according to the specifications
+                 given inside of the compound specifier.)
+   )
+   $(TR $(MULTIROW_CELL 1, $(I Delegate))
+        $(TD $(B 's'), $(B 'r'), compound)
+            $(TD Like the `.stringof` of this delegate treated as a string.
+
+                 $(I Please note: The implementation is currently buggy
+                 and it's use is discouraged.))
```
Comment (RazvanN7) [inline std/format/package.d:502]:
```suggestion
                 and its use is discouraged.))
```

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Mind the comments. Also, have you seen the actual rendering of this table? Maybe you can post it here as I'm afraid it may not look that well, given the amound of information present in it.

File: std/format/package.d (hunk 3 above)
Comment (berni44) [reply std/format/package.d:284]: This is decided by the hardware and can be queried and changed (on some computers) by using `std.math.hardware : FloatingPointControl`. If not available, default is round-to-nearest.

File: std/format/package.d (hunk 7 above)
Comment (berni44) [reply std/format/package.d:409]: I was quite unsure here. The German translation is the same for both. I'll change everything to "as".

File: std/format/package.d (hunk 3 above)
Comment (berni44) [reply std/format/package.d:284]: Suggestion: "Rounding of floating point numbers adheres to the rounding mode of the floating point unit, if available."

Comment (RazvanN7) [reply std/format/package.d]: Yes, that's great.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**